CIP-DappConnector/CIP-dapp-connector.md

```diff
@@ -0,0 +1,236 @@
+* UserDeclined - User declined to sign the data
+* InvalidFormat - If a wallet enforces data format requirements, this error signifies that the data did not conform to valid formats.
+
+### PaginateError
+
+```
+type PaginateError = {|
+    maxSize: number,
+|};
+```
+{maxSize} is the maximum size for pagination and if the dApp tries to request pages outside of this boundary this error is thrown.
+
+### TxSendError
+
+```
+TxSendErrorCode = {
+	Refused: 1,
+	Failure: 2,
+}
+type TxSendError = {
+	code: TxSendErrorCode,
+	info: String
+}
+```
+
+* Refused - Wallet refuses to send the tx (could be rate limiting)
+* Failure - Wallet could not send the tx
+
+### TxSignError
+
+```
+TxSignErrorCode = {
+	ProofGeneration: 1,
+	UserDeclined: 2,
+}
+type TxSignError = {
+	code: TxSignErrorCode,
+	info: String
+}
+```
+
+* ProofGeneration - User has accepted the transaction sign, but the wallet was unable to sign the transaction (e.g. not having some of the private keys)
+* UserDeclined - User declined to sign the transaction
+
+
+
+## Initial API
+
+In order to initiate communication from webpages to a user's Cardano wallet, the wallet must provide the following javascript API to the webpage. A shared, namespaced `cardano` object must be injected into the page if it did not exist already. Each wallet implementing this standard must then create a field in this object with a name unique to each wallet containing a `wallet` object with the following methods. The API is split into two stages to maintain the user's privacy, as the user will have to consent to `cardano.walletName.requestReadAccess()` in order for the dApp to read any information pertaining to the user's wallet.
+
+### wallet.requestReadAccess(): Promise\<API>
+
+Errors: APIError
+
+This is the entrypoint to start communication with the user's wallet. The wallet should request the user's permission to connect the web page to the user's wallet, and if permission has been granted, the full API will be returned to the dApp to use. The wallet can choose to maintain a whitelist to not necessarily ask the user's permission every time access is requested, but this behavior is up to the wallet and should be transparent to web pages using this API. If a wallet is already connected this function should not request access a second time, and instead just return the `API` object.
+
+### wallet.checkReadAccess(): Promise\<bool>
+
+Errors: APIError
+
+Returns true if the dApp is already connected to the user's wallet, or if requesting access would return true without user confirmation (e.g. the dApp is whitelisted), and false otherwise. If this function returns true, then any subsequent calls to `wallet.requestReadAccess()` during the current session should succeed and return the `API` object.
+
+### wallet.version(): String
+
+Errors: APIError
+
+Returns the version number of the API that the wallet supports.
+
+
+### wallet.name(): String
+
```

Comment:
Suggestion:

Change `wallet.requestReadAccess()` -> `wallet.enable()`
`wallet.checkReadAccess()` -> `wallet.isEnabled()`

I like to keep things simple and short and it would also align more with the `ethereum` API.
Maybe a `wallet.disable()` function would make sense as well (simply removing from whitelist)

What about having `wallet.version()`, `wallet.name()` as properties directly instead of funtions?

Reply:
Part of the reason why they had long names before was so we wouldn't inject code that would conflict with existing names, but now that it's something like `cardano.walletNameHere.requestReadAccess()` that sounds like a good idea to simplify it to just `cardano.walletNamehere.enable()`.

>What about having wallet.version(), wallet.name() as properties directly instead of funtions?

That could work too.

>Maybe a wallet.disable() function would make sense as well (simply removing from whitelist)

Are we imagining that `wallet.disable()` method to be for like a "sign out" button in the dApps? The way a whitelist is handled in the spec is left up to the wallet if and how they want to implement it, so whether this endpoint would simply end the session or remove seems like it would also be left up to the wallet. So I think the only guarantee we could make on this is that it disconnects the wallet from the page - subsequent calls to `enable()` could still reconnect you without user interaction if the wallet implements a whitelist and the site is whitelisted.

Reply:
Yeah I think the enable/disable functionality only really makes sense if the wallet has a whitelist, otherwise you don't really connect and disconnect and it's just a UI thing.

Reply:
Even without a whitelist it still makes some sense. You could have it simply disconnect the session between the dApp and the wallet, meaning no more access to the `api` endpoints until the dApp requests to connect again. This could be relevant for privacy-focused wallets/users who would like to control when the dApp can query their wallet I guess. Since the way i see it there are 3 possible connect/whitelist situations:

1. The whitelist is combined with the concept of being connected. This is what MetaMask and Yoroi's Ergo dApp connector use. To connect you need to effectively add the site to a whitelist, and then even if the user closes the browser/etc, the next time the dApp requests access it connects automatically. This is basically long-term session connections.
2. The whitelist exists and is separate from the concept of being connected. We had this as our initial design for the Ergo connector but switched to 1. for UX simplicity for the average user. So the whitelist just meant that calls to `wallet.requestReadAccess()` would resolve successfully without user input, and you could disconnect from a site without removing it from the whitelist.
3. There is no whitelist at all. You would have to accept connect requests every single website session.